src/ucs/datastruct/arbiter.c

```diff
@@ -67,28 +67,86 @@ void ucs_arbiter_group_head_desched(ucs_arbiter_t *arbiter,
     ucs_list_del(&head->list);
 }
 
-void ucs_arbiter_group_purge(ucs_arbiter_t *arbiter, ucs_arbiter_group_t *group,
+void ucs_arbiter_group_purge(ucs_arbiter_t *arbiter,
+                             ucs_arbiter_group_t *group,
                              ucs_arbiter_callback_t cb, void *cb_arg)
 {
-    ucs_arbiter_elem_t *tail = group->tail;
-    ucs_arbiter_elem_t *ptr, *next;
-    ucs_arbiter_elem_t *head;
+    ucs_arbiter_elem_t *tail       = group->tail;
+    ucs_arbiter_elem_t *next_group = NULL;
+    ucs_arbiter_elem_t *prev_group = NULL;
+    ucs_arbiter_elem_t *ptr, *next, *prev;
+    ucs_arbiter_elem_t *head, *orig_head;
+    ucs_arbiter_cb_result_t result;
+    int is_scheduled;
 
     if (tail == NULL) {
         return; /* Empty group */
     }
 
-    head = tail->next;
-    ucs_arbiter_group_head_desched(arbiter, head);
+    orig_head    = head = tail->next;
+    is_scheduled = (head->list.next != NULL);
+    next         = head;
+    prev         = tail;
+
+    if (is_scheduled) {
+        prev_group = ucs_list_prev(&head->list, ucs_arbiter_elem_t, list);
+        next_group = ucs_list_next(&head->list, ucs_arbiter_elem_t, list);
+    }
 
-    next = head;
     do {
-        ptr  = next;
-        next = ptr->next;
+        ptr       = next;
+        next      = ptr->next;
+        /* Can't touch the element if it gets removed. But it can be reused
+         * later as well, so it's next should be NULL. */
         ptr->next = NULL;
-        cb(arbiter, ptr, cb_arg);
+        result    = cb(arbiter, ptr, cb_arg);
+
+        if (result == UCS_ARBITER_CB_RESULT_REMOVE_ELEM) {
+            if (ptr == head) {
+                head = next;
+                if (ptr == tail) {
+                    /* Last element is being removed - mark group as empty */
+                    group->tail = NULL;
+                    /* Break here to keep ptr->next = NULL, otherwise ptr->next
+                       will be set to itself below */
+                    break;
+                }
+            } else if (ptr == tail) {
+                prev->next  = head;
+                group->tail = prev;
+            }
+            prev->next = next;
```

Comment:
do we set prev->next twice if (ptr == tail) ?
maybe in line 115 should ucs_assert(head == next) ?